torch/csrc/distributed/c10d/reducer.hpp

```diff
@@ -344,117 +345,104 @@ class TORCH_API Reducer {
+
+    // Variables whose gradients are held in this bucket
+    // We use refcounted tensors here so that we can easily unflatten the
+    // bucket contents into the participating variables after reduction has
+    // completed.
+    std::vector<at::Tensor> variables;
+
+    // Per-variable offset/length into the flattened `contents` tensor and
+    // the corresponding `GradBucket` instance for communication hooks
+    std::vector<size_t> offsets;
+    std::vector<size_t> lengths;
+
+    // Per-variable sizes slicing into the bucket contents
+    std::vector<c10::IntArrayRef> sizes_vec;
 
-    // Number of replicas to be marked done before this bucket is ready.
+    // Number of gradients left to be computed before the bucket is ready to
+    // be reduced
     size_t pending;
 
-    // Keep future work handle around DDP comm hook.
-    // If no hook is registered, a temporary vanilla allreduce hook will be
-    // used.
+    // Global indices of participating variables in the bucket
+    std::vector<size_t> variable_indices;
+
+    // Future work handle for DDP communication hook
+    // If no hook is registered, a temporary vanilla allreduce hook is used.
     c10::intrusive_ptr<at::ivalue::Future> future_work;
 
-    // If this bucket should expect a single sparse gradient.
-    // Implies: replicas[i].variables.size() == 1.
+    // If this bucket should expect a single sparse gradient
+    // If `true`, then this implies that `bucket.variables.size() == 1`.
     bool expect_sparse_gradient = false;
-    // "Limit" of cumulative parameter sizes that this bucket manages. It is
-    // actually a soft limit because we don't shard parameters across buckets
-    // so a single parameter may push it over the cap.
+
+    // "Limit" of the cumulative gradient sizes that this bucket manages
+    // It is actually a soft limit because we do not shard parameter across
+    // buckets, so a single parameter may push the bucket size over the limit.
     size_t bucket_size_limit;
+
+    // TODO(@pietern)
+    // Memory copies from gradient tensors into the bucket are potentially
+    // done on different CUDA streams. We record an event for every copy
+    // so that we can synchronize with them prior to kicking off the reduction.
+    // std::vector<at::cuda::CUDAEvent> events;
+
   };
 
   std::vector<Bucket> buckets_;
 
-  // A variable locator locates a particular variable in the bucket
-  // structure. The `bucket_index` field points to the bucket in the `buckets_`
-  // vector. The `intra_bucket_index` field points to the index of the variable
-  // in any of the vector fields in the bucket replica.
+  // A variable locator locates a particular variable in the reducer's buckets
   struct VariableLocator {
-    // Index into the `buckets_` variable.
+    // Index of the bucket containing the variable in the `buckets_` vector
     size_t bucket_index;
-    // Index of parameter in single bucket replica.
+    // Index of the variable in the bucket
```

Comment:
update this to mention it is the index of `variables`, `bucket_views_in`, etc structures?